### PR TITLE
HIVE-26157: Change Iceberg storage handler authz URI to metadata location

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
Change Iceberg storage handler autz URI from `iceberg://dbName/tableName` format to `iceberg://metadataLocation`

### Why are the changes needed?
It is possible to set the metadata pointers of table A to point to table B, and therefore you could read table B's data via querying table A.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manual test, unit test

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 1]
```diff
@@ -450,7 +452,9 @@ public boolean isValidMetadataTable(String metaTableName) {
   public URI getURIForAuth(org.apache.hadoop.hive.metastore.api.Table hmsTable) throws URISyntaxException {
     String dbName = hmsTable.getDbName();
     String tableName = hmsTable.getTableName();
-    return new URI(ICEBERG_URI_PREFIX + dbName + "/" + tableName);
+    Table table = IcebergTableUtil.getTable(conf, hmsTable);
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:455]: We ran into problems with the approach of loading the Iceberg table here before. The problem is that this method can be called to authorize CREATE TABLE commands as well, at which point the iceberg table does not exist yet, so this will lead to NPE.

If the table object is null, then maybe we can use the hmsTable.getSd().location() + "/metadata"? I'm not sure though, just thinking out loud

Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:455]: Giving back `hmsTable.getSd().location() + "/metadata"` seems reasonable to me in such cases

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 2]
```diff
@@ -104,7 +106,7 @@
 public class HiveIcebergStorageHandler implements HiveStoragePredicateHandler, HiveStorageHandler {
   private static final Logger LOG = LoggerFactory.getLogger(HiveIcebergStorageHandler.class);
 
-  private static final String ICEBERG_URI_PREFIX = "iceberg://";
+  private static final String ICEBERG_URI_PREFIX = "iceberg:/";
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:109]: Was this change necessary? I think it's supposed to work like a 'protocol' that's why there is the double-slash. I think kafka and druid did the same thing (i.e. `druid://` and `kafka://`)

**Comment by lcspinter** [comment]
PR #3234 will resolve the test failures.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 1 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:462]: What happens if we provide the `metadata_location` in the table properties when creating the table?
Or do we plan to handle this in a different PR?

Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:462]: Good point, I will add a change to address this use case as well.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 3]
```diff
@@ -456,9 +458,19 @@ public boolean isValidMetadataTable(String metaTableName) {
   public URI getURIForAuth(org.apache.hadoop.hive.metastore.api.Table hmsTable) throws URISyntaxException {
     String dbName = hmsTable.getDbName();
     String tableName = hmsTable.getTableName();
-    return new URI(ICEBERG_URI_PREFIX + dbName + "/" + tableName);
+    Optional<String> locationProperty = SessionStateUtil.getProperty(conf, hive_metastoreConstants.META_TABLE_LOCATION);
+    if (locationProperty.isPresent()) {
+      // this property is set during the create operation before the hive table was created
+      // we are returning a dummy iceberg metadata file
+      return new URI(ICEBERG_URI_PREFIX + locationProperty.get() + "/metadata/dummy.metadata.json");
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:465]: Do we always set the location?
Would it be worthwhile to check this?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 4]
```diff
@@ -13688,6 +13688,22 @@ ASTNode analyzeCreateTable(
       // outputs is empty, which means this create table happens in the current
       // database.
       rootTasks.add(TaskFactory.get(new DDLWork(getInputs(), getOutputs(), crtTblDesc)));
+      String tblLocation = null;
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13691]: What happens with the location if the Transaltion kicks in?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 3 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:465]: yes, we are setting it. That is why I added that additional code change to the SemanticAnalyzer

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java [hunk 5]
```diff
@@ -1411,15 +1411,32 @@ public void testCommandsWithPartitionClauseThrow() {
...
     URI uriForAuth = storageHandler.getURIForAuth(hmsTable);
 
-    Assert.assertEquals("iceberg://" + hmsTable.getDbName() + "/" + hmsTable.getTableName(), uriForAuth.toString());
+    Assert.assertEquals("iceberg:/" + URI.create(((BaseTable) table).operations().current()
+            .metadataFileLocation()).getPath(), uriForAuth.toString());
+
+  }
+
+  @Test
+  public void testAutzURIWithAuthEnabled() throws TException, InterruptedException, URISyntaxException {
```
[the remark this answers is not in the document]
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:1428]: I'm not sure how can we test this. We would need to halt the create table operation before the iceberg table is created. Do you have some suggestions?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 4 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:13691]: There is a `Translator` layer in HMS which can change some properties fo the table (not sure which ones), but I remember that this `Translator` was responsible to changing location for the external tables. Does this change able to handle this `Translator`?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:1428]: Can we check which URL is checked for security on table creation? Maybe with a Mock security handler?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 3 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:465]: So a `Precondition` check could be ok here?

Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:465]: sure, that can be done

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 6]
```diff
@@ -457,9 +459,24 @@ public boolean isValidMetadataTable(String metaTableName) {
   public URI getURIForAuth(org.apache.hadoop.hive.metastore.api.Table hmsTable) throws URISyntaxException {
     String dbName = hmsTable.getDbName();
     String tableName = hmsTable.getTableName();
-    return new URI(ICEBERG_URI_PREFIX + dbName + "/" + tableName);
+    StringBuilder authURI = new StringBuilder(ICEBERG_URI_PREFIX).append(dbName).append("/").append(tableName)
+        .append("?snapshot=");
+    Optional<String> locationProperty = SessionStateUtil.getProperty(conf, hive_metastoreConstants.META_TABLE_LOCATION);
+    if (locationProperty.isPresent()) {
+      Preconditions.checkArgument(locationProperty.get() != null,
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:466]: How can this happen?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 7]
```diff
@@ -457,9 +459,24 @@ public boolean isValidMetadataTable(String metaTableName) {
...
+          "Table location is not set " + "in SessionState. Authorization URI cannot be supplied.");
+      // this property is set during the create operation before the hive table was created
+      // we are returning a dummy iceberg metadata file
+      authURI.append(URI.create(locationProperty.get()).getPath()).append("/metadata/dummy.metadata.json");
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:470]: Maybe log at least on debug level?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 8]
```diff
@@ -457,9 +459,24 @@ public boolean isValidMetadataTable(String metaTableName) {
...
+      return new URI(HiveConf.EncoderDecoderFactory.URL_ENCODER_DECODER.encode(authURI.toString()));
+    }
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:472]: nit: maybe use an `else`? I usually try to avoid using return in the middle of a method and using the return statement in place of else statements.
```
if () {
    return...
} else {
    return...
}
```
For me this is easier to read.

What do you think?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 6 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:466]: I believe this shouldn't happen anymore. Before my fix in `SessionStateUtil#getQueryState()` could have returned null values.

Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:466]: Shall we remove it then?

Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:466]: I could remove it, but I think it doesn't hurt if we have an additional check validating the location.

**Review by pvary — APPROVED** [review]
+1 pending tests

**Review by pvary — APPROVED** [review]
+1 pending tests
